### ISSUE TITLE
Having problems using this

### ISSUE THREAD

**Issue by Daenara** [jmc265/node-red-contrib-rpi-433#3]
Hello,
 
I am struggling getting this to work for me. I can use `./RFSniffer 0 23` if I navigate to the rpi-433-v2/build directory and it receives the codes from my remote. I can use `./RFSniffer` after I add in my pin in the RFSniffer.cpp and fix a bug that always overwrites the pin with 0 and I receive my codes. But if I use your node-red module I don't receive anything. Sending doesn't work either, but I did not try to debug that right now because I don't need it for my project as much as I need receiving.

I copied the example app from the rpi-433-v2 readme, added my pin and tried using that and I don't receive anything. I added a few debug outputs in the Sniffer.js to check that the correct pin ends up in there and that the path to the RFSniffer executable is correct, both is the case. I saw then that it tries to add the pin with a --pin option, which is not even an option for RFSniffer as far as I can see in the code of that. I edited that section to not add any options when executing RFSniffer because I added my pin into the script manually and it still didn't do anything. Calling `/usr/src/node-red/node_modules/rpi-433-v2/build/RFSniffer` manually works for me and that is literally what the spawn command in the Sniffer.js script should call after removing the --pin argument it had.

So I am concluding with my almost non-existent js knowledge that the rpi-433-v2 package is broken somehow and I am wondering how you were able to use it for this project because I assume you tested it when you switched to the v2 project.

I would take this over to their github page but they don't have an issue button for some reason and the pull request was open for a while over there also. I don't know enough js to fix anything in there, I am not even sure what exactly is broken (besides the pin override in the cpp files and the --pin argument that the cpp files don't take) but I hope you have some insight in this (or anyone just stumbling over this post).

This has been annoying me for the past 3 weeks now, I hope someone can help.

Daenara

EDIT: Just found out, the sending part works fine, I clicked the node just testing for the 500th time and my lamp went off. Seems like I never tested the sending part while my lamp was on so I thought it didn't work also because the receiver did not receive the code sent. So just the sniffer is broken.

**Comment by jmc265** [comment]
Hi

I am glad you got the sender node to work. The sniffer shouldn't really operate any differently so I am not sure what would be broken about the sniffer side of things. I have tested the v2 package when I created these nodes. I am not using them currently, but will be using them again soon so I can take another look then, but I expect them to work without any modification. 

It appears as though the v2 github repo has been removed, so only the npm page is left: https://www.npmjs.com/package/rpi-433-v2 which is strange, so you can't open an issue there. You can try using this package before I merged in v2 by installing verson 1.0.0 `npm install node-red-contrib-rpi-433@1.0.0`. Let me know how you get on.

**Comment by Daenara** [comment]
I just tried the old version, here is what happened:

```
+ npm install node-red-contrib-rpi-433@1.0.0

> rpi-433@2.1.0 install /usr/src/node-red/node_modules/rpi-433
> mkdir ./build && cd ./src && /usr/bin/make

g++    -c -o RCSwitch.o RCSwitch.cpp
RCSwitch.cpp: In member function 'char* RCSwitch::getCodeWordB(int, int, boolean)':
RCSwitch.cpp:196:12: error: invalid conversion from 'char' to 'char*' [-fpermissive]
     return '\0';
            ^~~~
RCSwitch.cpp: In member function 'char* RCSwitch::getCodeWordA(char*, int, boolean)':
RCSwitch.cpp:232:14: error: invalid conversion from 'char' to 'char*' [-fpermissive]
       return '\0';
              ^~~~
RCSwitch.cpp:241:14: error: invalid conversion from 'char' to 'char*' [-fpermissive]
       return '\0';
              ^~~~
RCSwitch.cpp: In member function 'char* RCSwitch::getCodeWordC(char, int, int, boolean)':
RCSwitch.cpp:269:12: error: invalid conversion from 'char' to 'char*' [-fpermissive]
     return '\0';
            ^~~~
RCSwitch.cpp: In static member function 'static bool RCSwitch::receiveProtocol1(unsigned int)':
RCSwitch.cpp:524:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
RCSwitch.cpp: In static member function 'static bool RCSwitch::receiveProtocol2(unsigned int)':
RCSwitch.cpp:559:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
make: *** [<builtin>: RCSwitch.o] Error 1
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! rpi-433@2.1.0 install: `mkdir ./build && cd ./src && /usr/bin/make`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the rpi-433@2.1.0 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2020-07-24T18_46_09_843Z-debug.log

```

For some reason the version of rpi-433 does not compile. It seems that this will continue to annoy me for quite some time. I even did a clean install hoping that would help.

My main problem is that the compiled c code works but the javascropt code doesn't because I literally never worked with javascript before so finding what is wrong in there is hard.

**Comment by jmc265** [comment]
Ah I remember I had an issue compiling that version of rpi-433 as well. You have to make sure you set the `fpermissive` flag for GCC, although I can't remember how to do that off the top of my head. There was an env var you can export with GCC options in it...

As I said, I am soon going to start using the sniffer node in this repo to listen to my wireless doorbell. I will get back to you when/if I can get it to work, or release a fix. But last I was using it a couple of months ago and it was working fine.

**Comment by Daenara** [comment]
Thanks for that tip. It took me a while to find the right variable but the old version runs perfectly. I can finally retire my hacked python script and use my custom node red docker instead.

**Comment by jmc265** [comment]
Ok, I have fixed it!

It turns out the rpi-433-v2 repo used a version of RPSniffer.cpp which did not flush its output to stdout. So there was nothing wrong with the JS, it was just waiting for output that never got flushed to it. I have created [rpi-433-v3](https://github.com/jmc265/rpi-433-v3/tree/master/src) and published that. I have also updated node-red-contrib-rpi-433 to version 1.2.3 which uses this new -v3 library. I am closing this issue now, but let me know if you get it working.

**Comment by Daenara** [comment]
okay, that is not something I could have found on my own without actually understanding what the javascript does. I just tested the new version and it works. Thanks for the quick fix and the help along the way.

**Comment by Daenara** [comment]
I just noticed that RFSniffer uses 105% of my cpu, can you check if it does it for you, also or if I messed something up somewhere?

**Comment by jmc265** [comment]
Yep, I am finding the same, rfsniffer takes up 100% of one of the cores of the CPU.

I've just taken a look into it and I think I may have fixed it with this: https://github.com/jmc265/rpi-433-v3/commit/8be88e6d31b26db38dc0ea623eb50ea335c5696b

I have updated this package to point to the new version and you should be able to install with `npm install node-red-contrib-rpi-433@1.3.0`. Let me know if that helps.

**Comment by Daenara** [comment]
If I had checked here earlier instead of just reading the mail notification I could have been spared an hour of checking every fork of the original 433Utils for any updates that seem useful. 

It works, not sure if my remote is messing with me or if I really managed to hit the button during the sleep interval 3 times straight but I think that the remote did stuff like that even when I used it to actually switch a rf outlet.

CPU usage is down to about 25%, still the highest on my raspi but it is down to a manageable level.  Some ppl have gotten it down to 11% with an even longer sleep but I don't think that would be as responsive anymore.

Thank you for the quick fix
